### PR TITLE
rust: types: Add `try_from_foreign()` method

### LINKED ISSUE #1057: `try_from_foreign()`
Currently `ForeignOwnable::from_foreign()` only works for non-null pointers for the existing `impl`s (e.g. `Box`, `Arc`). It may create a few duplicate code like:

```rust
// `p` is a maybe null pointer
if p.is_null() {
    None
} else {
     Some(unsafe { T::from_foreign(p) })
}
```

see [this patch](https://lore.kernel.org/rust-for-linux/20240116151728.370238-1-mcanal@igalia.com/) for example.

It's better if we have a function: `*const void` -> `Option<ForeignOwnable>`, which can handle null pointer cases. There are at least two ways I can think of:

1. a `try_from_foreign()` function in `ForeignOwnable` trait:
```rust
pub trait ForeignOwnable {
    unsafe fn try_from_foreign(ptr: *const void) -> Option<Self> {
        if ptr.is_null() {
            None
        } else {
            Some(Self::from_foreign(ptr))
        }
    }
}
```
2. `impl ForeignOwnable` for `Option<T: ForeignOwnable>`:

```rust
impl<T: ForeignOwnable> ForeignOwnable for Option<T> {
    type Borrowed<'a> = Option<T::Borrowed<'a>>;
    unsafe fn borrow<'a>(ptr: *const void) -> Self::Borrowed<'a> {
        if ptr.is_null() {
            None
        } else {
            Some(T::borrow(ptr))
        }
     }
    unsafe fn from_foreign(ptr: *const void) -> Self {
        if ptr.is_null() {
            None
        } else {
            Some(T::from_foreign(ptr))
        }
     }
}
```

Of course, there could be better ideas ;-)

### PR DESCRIPTION
**Pull request by obeis**
Close #1057

### CODE REVIEW

File: rust/kernel/types.rs [hunk 1]
```diff
@@ -46,6 +46,18 @@ pub trait ForeignOwnable: Sized {
     /// Additionally, all instances (if any) of values returned by [`ForeignOwnable::borrow`] for
     /// this object must have been dropped.
     unsafe fn from_foreign(ptr: *const core::ffi::c_void) -> Self;
+
+    /// Converts a foreign-owned object back to a Rust-owned one.
+    ///
+    /// similar to `from_foreign`, but if `ptr` is null, it will return `None`.
```
Comment (ojeda) [inline rust/kernel/types.rs]: Perhaps:

```suggestion
    /// Tries to convert a foreign-owned object back to a Rust-owned one.
    ///
    /// Like [`from_foreign`], but returns [`None`] if `ptr` is null.
```

(I didn't check if the intra-doc links would work as-is)

**Comment by ojeda** [comment]
Thanks!

The commit should be signed (in the `Signed-off-by` sense, please see https://docs.kernel.org/process/submitting-patches.html#developer-s-certificate-of-origin-1-1) and should have a `Link:` tag to the issue.

The commit message could perhaps also show how a user would look like (like one of the ones linked in the issue).

Apart from that, it seems OK to submit it to the mailing list -- please see https://rust-for-linux.com/contributing.

**Comment by obeis** [comment]
I have sent it to the mailing list.

https://lore.kernel.org/rust-for-linux/0100018d31fe5238-538d9c26-64a3-49cb-ac76-22362f73d80f-000000@email.amazonses.com

**Comment by charmitro** [comment]
Looks good. Is there code in the tree that can use this right away or we'd have to wait for patches to get merged?

**Comment by ojeda** [comment]
`try_from_foreign` (option 1 in the linked issue) is now in `rust-next`.

**Comment by obeis** [comment]
Done:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=5bc818419a551ab1fcdbff3c5d8b6043527d90dd

**Pull request closed without merging**